clang/runtime/dpct-rt/include/dpct/dpl_extras/algorithm.h

```diff
@@ -937,18 +966,29 @@ stable_partition(Policy &&policy, Iter1 first, Iter1 last, Iter2 mask, Pred p) {
                        std::random_access_iterator_tag>::value,
       "Iterators passed to algorithms must be random-access iterators.");
   typedef typename std::decay<Policy>::type policy_type;
-  internal::__buffer<typename std::iterator_traits<Iter1>::value_type> _tmp(
-      std::distance(first, last));
+  auto _partition_call1 =
+      internal::make_wrapped_policy<internal::partition_call1>(policy);
+  auto _copy_call1 =
+      internal::make_wrapped_policy<internal::copy_call1>(policy);
+  auto _copy_call2 = internal::make_wrapped_policy<internal::copy_call2>(
+      std::forward<Policy>(policy));
 
-  std::copy(policy, mask, mask + std::distance(first, last), _tmp.get());
+  auto _n = std::distance(first, last);
+  internal::__buffer<typename std::iterator_traits<Iter1>::value_type> _tmp(_n);
 
-  auto ret_val =
-      std::stable_partition(std::forward<Policy>(policy),
-                            oneapi::dpl::make_zip_iterator(first, _tmp.get()),
-                            oneapi::dpl::make_zip_iterator(
-                                last, _tmp.get() + std::distance(first, last)),
-                            internal::predicate_key_fun<Pred>(p));
-  return std::get<0>(ret_val.base());
+  auto _tmp_first = _tmp.get();
+  auto _tmp_last = _tmp_first + _n;
+  auto _tmp_last_rev = std::reverse_iterator(_tmp_last);
+
+  auto _end_pair =
+      stable_partition_copy(std::move(_partition_call1), first, last, mask,
+                            _tmp_first, _tmp_last_rev, p);
+  auto _first_n = std::distance(_tmp_first, std::get<0>(_end_pair));
+  std::copy(std::move(_copy_call1), _tmp_first, std::get<0>(_end_pair), first);
```

Comment:
How about using this function's return value instead of `first + _first_n` to avoid computing the distance?

Reply:
Thanks, this is a good suggestion.  Fixed.